akka-stream/src/main/scala/akka/stream/impl/fusing/FoldWithin.scala

```diff
@@ -48,50 +49,52 @@ class FoldWith[In, Agg, Out](
   override def createLogic(inheritedAttributes: Attributes): GraphStageLogic =
     new TimerGraphStageLogic(shape) with InHandler with OutHandler {
 
-      // mutable state to keep track of the aggregator status to coordinate the flow
-      // input/output handler callbacks are guaranteed to execute without concurrency
-      // https://doc.akka.io/docs/akka/current/stream/stream-customize.html#thread-safety-of-custom-operators
       private[this] var aggregator: Agg = null.asInstanceOf[Agg]
 
+      private val buffer = mutable.Queue[Out]()
+
+      private def bufferFull: Boolean = buffer.size >= maxBufferSize
```

Comment:
I don't think the additional complexity added by potentially allowing multiple completed aggregates/groups to queue up on backpressure is a good idea. That can anyway be achieved by composing this operator with a buffer operator.

What I was talking about earlier was about keeping pulling from upstream until the group is complete even if downstream applies backpressure and propagating the backpressure only once the group is complete and cannot be pushed downstream.

Reply:
ok, removed the buffer. I figured out what was the reason to add the buffer.
Previously, back pressure was only based on isAvailable(out). Since isAvailable(out) is initially false without the first downstream request, this does not work for the initial condition to eagerly pull until first emit.
So I have to add another `hasEmitted` flag so it should be equivalent to bufferSize=1 now.

Reply:
Actually, the eagerly pull logic has to rely on an internal buffer even its size is just 1.
Otherwise, it can only react to onPull from downstream only when there is more requests than emits, which is too late for a truly eagerly pull.
The boolean flag only works for the initial eagerly pull. The extra emit has to be prepared ahead of time and served onPull when available.
I don't see there is any simpler logic other than using a general buffer to serve that purpose.
This is based on observation of the test case.

Reply:
If composing with the buffer operator as you suggested can always generate the demand from downstream, then there is no point to eagerly pull from within this stage.

Reply:
removed internal buffer and eager pull as a whole package, as the behavior is equivalent to composing with the Buffer operator with the backpressure OverflowStrategy

Reply:
Eager pull initial pull makes sense since this stage already is somewhat detaching upstream and downstream demand in the same way buffer does, you can add to aggregate even if there is no demand from downstream yet, and think of the aggregate state as that 1 element buffer which you don't send downstream if there is no demand. Eagerly pulling in preStart would allow for the operator to start creating an aggregate, and potentially have it ready once downstream demand is seen if downstream backpressures initially.

I think this is how all the other aggregating operators work, so that is why I am insistent on doing that here as well, to align.